bluesky_kafka/__init__.py

```diff
@@ -1,4 +1,287 @@
+        )
+
+    def flush(self):
+        """
+        Flush all buffered messages to the broker(s).
+        """
+        self.producer.flush()
+
+
+class RemoteDispatcher(Dispatcher):
+    """
+    Dispatch documents received over the network from a Kafka server.
+
+    There is no default configuration. A reasonable configuration for production is
+        consumer_config={
+            "auto.offset.reset": "latest"
+        }
+
+    Parameters
+    ----------
+    topics: list
+        List of topics as strings such as ["topic-1", "topic-2"]
+    bootstrap_servers : str
+        Comma-delimited list of Kafka server addresses as a string such as ``'127.0.0.1:9092'``
+    group_id: str
+        Required string identifier for Kafka Consumer group
+    consumer_config: dict
+        Override default configuration or specify additional configuration
+        options to confluent_kafka.Consumer.
+    polling_duration: float
+        Time in seconds to wait for a message before running function work_while_waiting.
+        Default is 0.05.
+    deserializer: function, optional
+        optional function to deserialize data. Default is pickle.loads.
+
+    Example
+    -------
+
+    Print all documents generated by remote RunEngines.
+
+    >>> d = RemoteDispatcher(
+    >>>         topics=["abc.def", "ghi.jkl"],
+    >>>         bootstrap_servers='localhost:9092',
+    >>>         group_id="xyz",
+    >>>         consumer_config={
+    >>>             "auto.offset.reset": "latest"
+    >>>         }
+    >>>    )
+    >>> d.subscribe(print)
+    >>> d.start()  # runs until interrupted
+    """
+
+    def __init__(
+        self,
+        topics,
+        bootstrap_servers,
+        group_id,
+        consumer_config=None,
+        polling_duration=0.05,
+        deserializer=pickle.loads,
+    ):
+        super().__init__()
+
+        self.polling_duration = polling_duration
+        self._deserializer = deserializer
+
+        self.consumer_config = dict()
+        if consumer_config is not None:
+            self.consumer_config.update(consumer_config)
+
+        if "group.id" in self.consumer_config:
+            raise ValueError(
+                "do not specify 'group.id' in consumer_config, use only the 'group_id' argument"
+            )
+        else:
+            self.consumer_config["group.id"] = group_id
+
+        if "bootstrap.servers" in self.consumer_config:
+            self.consumer_config["bootstrap.servers"] = ",".join(
+                [bootstrap_servers, self.consumer_config["bootstrap.servers"]]
+            )
+        else:
+            self.consumer_config["bootstrap.servers"] = bootstrap_servers
+
+
+        logger.info(
+            "starting RemoteDispatcher with Kafka Consumer configuration:\n%s",
+            consumer_config,
+        )
+        logger.info("subscribing to Kafka topic(s): %s", topics)
+
+        self.consumer = Consumer(self.consumer_config)
+        self.consumer.subscribe(topics=topics)
```

Comment:
It might be convenient to be able to inspect which `topics` a `RemoteDispatcher` instance is subscribed to. Do you think it's worth stashing this in instance state? (If so, it should be exposed through a read-only property since changing it after the fact would have no effect on the underlying `Consumer`.)

Reply:
Yes - stashing.

Reply:
Done!